src/Halogen/ECharts.purs

```diff
@@ -1,132 +1,287 @@
-
-  dataHalogenEChartsOptions :: forall i. EncodedOptions -> A.Attr i
-  dataHalogenEChartsOptions = A.attr $ A.attributeName "halogen-echarts-options"
-  
-newtype EncodedOptions = EncodedOptions Json
-  
-instance optionIsAttribute :: A.IsAttribute EncodedOptions where
-  toAttrString _ _ = "EncodedOptions"
-      
-foreign import getOptions
-  "function getOptions(node) {\
-  \  return function() {\
-  \    return node['halogen-echarts-options'];\
-  \  };\
-  \}" :: forall eff. HTMLElement -> Eff (dom :: DOM | eff) EncodedOptions
-      
-foreign import setOptions
-  "function setOptions(option, chart) {\
-  \  return function() {\
-  \    chart.setOption(option);\
-  \  };\
-  \}" :: forall eff. Fn2 EncodedOptions EC.EChart (Eff (ECEffects eff) Unit)
-      
-updateOptions :: forall eff. HTMLElement -> EC.EChart -> Eff (ECEffects eff) Unit
-updateOptions node ec = void do
-  opts <- getOptions node
-  runFn2 setOptions opts ec
\ No newline at end of file
+  rearrange'
+    :: Ref (StrMap {inst :: String, el :: Element})
+    -> Int -> Eff (dom :: DOM, ref :: REF|e) Unit
+  rearrange' next n = do
+    -- Get all components
+    echartsContainers <- window
+                         >>= document
+                         >>= querySelectorAll "[data-echarts-key]"
+                         <<< htmlDocumentToParentNode
+    -- Count them
+    count <- length echartsContainers
+    if n >= count
+      -- we've processed all components, let's remove unneeded keys from memo
+      -- by flipping refs content
+      then readRef next >>= writeRef memo
+      else do
+      -- move one component
+      moveOne next n echartsContainers
+      -- continue
+      rearrange' next $ n + one
+
+  moveOne :: Ref (StrMap {inst :: String, el :: Element})
+          -> Int -> NodeList -> Eff (dom :: DOM, ref :: REF|e) Unit
+  moveOne next ix containers = void $ runMaybeT do
+      -- try to take component element
+      container <- MaybeT $ map toMaybe $ item ix containers
+      -- try to find chart element
+      lift do mbel <- map (toMaybe >>> map elementToNode)
+                        $ querySelector "div"
+                        $ nodeToParentNode container
+              -- and remove if it exists
+              maybe (pure unit) (void <<< flip removeChild container) mbel
+      datas <- lift $ dataset container
+      -- get key to take memoized el and _echarts_instance_
+      key <- maybe empty pure $ lookup "echartsKey" datas
+      m <- lift $ readRef memo
+      r <- maybe empty pure $ lookup key m
+      -- append element, set attribute
+      lift do
+        appendChild (elementToNode r.el) container
+        setAttribute "_echarts_instance_" r.inst $ nodeToElement container
+        -- Everything works, this el/chart is exists, let's write it to new ref
+        modifyRef next $ insert key r
+
+  nodeToParentNode :: Node -> ParentNode
+  nodeToParentNode = unsafeCoerce
```

Comment:
For each use of `unsafeCoerce`, would you mind adding a brief comment explaining why it is safe? Just in case someone looks at this later and wants to understand what is going on. Thanks!!